service/matching/matching_engine.go

```diff
@@ -837,6 +837,116 @@ func (e *matchingEngineImpl) listTaskQueuePartitions(request *matchingservice.Li
+	if err != nil {
+		return nil, err
+	}
+
+	tqMgr, err := e.getTaskQueuePartitionManager(ctx, taskQueue, normalStickyInfo, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// we don't set updateOptions.TaskQueueLimitPerBuildId, because the Versioning Rule limits will be checked separately
+	// we don't set updateOptions.KnownVersion, because we handle external API call ordering with conflictToken
+	updateOptions := UserDataUpdateOptions{}
+	cT := req.GetConflictToken()
+
+	err = tqMgr.UpdateUserData(ctx, updateOptions, func(data *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, bool, error) {
+		clk := data.GetClock()
+		if clk == nil {
+clk = hlc.Zero(e.clusterMeta.GetClusterID())
+		} else {
+			prevCT, err := clk.Marshal()
+			if err != nil {
+				return nil, false, err
+			}
+			if !bytes.Equal(cT, prevCT) {
+				return nil, false, serviceerror.NewFailedPrecondition(
+					fmt.Sprintf("provided conflict token %v does not match existing one %v", cT, prevCT),
+				)
+			}
+		}
+		updatedClock := hlc.Next(clk, e.timeSource)
+		hadUnfiltered := hasUnfiltered(data.GetVersioningData().GetAssignmentRules())
+		var versioningData *persistencespb.VersioningData
+		switch req.GetOperation().(type) {
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertAssignmentRule:
+			versioningData, err = InsertAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertAssignmentRule(),
+				e.config.VersionAssignmentRuleLimitPerQueue(ns.Name().String()),
+				hadUnfiltered,
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceAssignmentRule:
+			versioningData, err = ReplaceAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceAssignmentRule(),
+				hadUnfiltered,
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteAssignmentRule:
+			versioningData, err = DeleteAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteAssignmentRule(),
+				hadUnfiltered,
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertCompatibleRedirectRule:
+			versioningData, err = InsertCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertCompatibleRedirectRule(),
+				e.config.VersionRedirectRuleLimitPerQueue(ns.Name().String()),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleRedirectRule:
+			versioningData, err = ReplaceCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceCompatibleRedirectRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleRedirectRule:
+			versioningData, err = DeleteCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteCompatibleRedirectRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId_:
```

Comment:
the trailing underscore, is it new or you just use type already defined?

Reply:
there are two CommitBuildId types defined, one with underscore, and one without. The one without gives a compilation error, so I'm using the one with. However, the implementation of that case is not one of my tasks this sprint, so I'm leaving any thoughts about that to when I'm working on the ticket for `CommitBuildId`